pgjdbc/src/test/java/org/postgresql/test/jdbc2/TimezoneCachingTest.java

```diff
@@ -0,0 +1,411 @@
+          ts2, rs.getTimestamp(2));
+      TimeZone.setDefault(tz1);
+      assertEquals(
+          "Cache was initialized to tz2, so timestamps cannot be equal",
+          ts2, rs.getTimestamp(2));
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      assertEquals(
+          "Explicit tz2 calendar, so timestamps cannot be equal",
+          ts2, rs.getTimestamp(2, c2));
+      assertEquals(
+          "Current TZ is tz1, empty cache to be initialized to tz1 => retrieve in tz1, timestamps must be equal",
+          ts1, rs.getTimestamp(2));
+      assertEquals(
+          "Explicit tz2 calendar, so timestamps cannot be equal",
+          ts2, rs.getTimestamp(2, c2));
+      assertEquals(
+          "Explicit tz2 calendar, so timestamps must be equal",
+          ts1, rs.getTimestamp(2, c1));
+      rs.close();
+    } finally {
+      TimeZone.setDefault(null);
+      TestUtil.closeQuietly(rs);
+      TestUtil.closeQuietly(pstmt);
+      TestUtil.closeQuietly(stmt);
+    }
+  }
+
+  private void checkTimestamp(String checkText, Statement stmt, Timestamp ts, TimeZone tz)
+      throws SQLException {
+    TimeZone prevTz = TimeZone.getDefault();
+    TimeZone.setDefault(tz);
+    ResultSet rs = stmt.executeQuery("SELECT col2 FROM testtz");
+    rs.next();
+    Timestamp dbTs = rs.getTimestamp(1);
+    rs.close();
+    TimeZone.setDefault(prevTz);
+    assertEquals(checkText, ts, dbTs);
+  }
+
+  private TimeZone getTimeZoneCache(Object stmt) {
+    try {
+      Field defaultTimeZoneField = stmt.getClass().getDeclaredField("defaultTimeZone");
+      defaultTimeZoneField.setAccessible(true);
+      return (TimeZone) defaultTimeZoneField.get(stmt);
+    } catch (Exception e) {
+    }
+    return null;
+  }
+
+  public TimezoneCachingTest(String name) {
+    super(name);
+    try {
+      Class.forName("org.postgresql.Driver");
+    } catch (Exception ex) {
+    }
+  }
+
+  /* Set up the fixture for this test case: a connection to a database with
+  a table for this test. */
+  protected void setUp() throws Exception {
+    super.setUp();
+    Statement stmt = con.createStatement();
+
+    /* Drop the test table if it already exists for some reason. It is
+    not an error if it doesn't exist. */
+    TestUtil.createTable(con, "testtz", "col1 INTEGER, col2 TIMESTAMP");
+
+    stmt.close();
+
+    /* Generally recommended with batch updates. By default we run all
+    tests in this test case with autoCommit disabled. */
+    con.setAutoCommit(false);
+  }
+
+  // Tear down the fixture for this test case.
+  protected void tearDown() throws SQLException {
+    con.setAutoCommit(true);
+
+    TestUtil.dropTable(con, "testtz");
+    super.tearDown();
+  }
+
+  @Override
+  protected void updateProperties(Properties props) {
+    props.setProperty(PGProperty.REWRITE_BATCHED_INSERTS.getName(),
+        Boolean.TRUE.toString());
```

Comment:
Is it required for the test? Does it explore something different in rewrite=true mode?